src/main/webapp/app/modules/canvas/visualization/visualization-modal/visualization-edit-modal/visualization-edit-modal-popup.tsx

```diff
@@ -0,0 +1,151 @@
+import {
+  getEntity as getVisualmetadataEntity,
+  updateEntity as updateVisualmetadataEntity,
+} from 'app/entities/visualmetadata/visualmetadata.reducer';
+import { getDatasourcesFeaturesEntities as getfeatureEntities } from 'app/entities/feature/feature.reducer';
+import { getEntity as getViewEntity } from 'app/entities/views/views.reducer';
+import { VisualMetadataContainerUpdate } from '../../util/visualmetadata-container.service';
+import { setVisual } from '../../util/VisualDispatchService';
+import {  renderVisualization, ValidateFields } from '../../util/visualization-render-utils';
+import { VisualWrap } from '../../util/visualmetadata-wrapper';
+import { forwardCall } from 'app/shared/websocket/proxy-websocket.service';
+import { subscribeWebSocket } from 'app/shared/websocket/stomp-client.service';
+
+export interface IVisualizationEditModalProps extends StateProps, DispatchProps {
+  id: number;
+  viewId: number;
+  visualizationId: string;
+  setOpen: (isOpen: boolean) => void;
+}
+
+export const VisualizationEditModal = (props: IVisualizationEditModalProps) => {
+  const [getData, setData] = useState<any>();
+  const dialog = useDialogContainer();
+  const visualizationId = props.visualizationId;
+  const viewId = props.viewId;
+
+  const handleClose = () => {
+    props.setOpen(false);
+    setVisual(props.visualmetadataEntity);
+    dialog.dismiss();
+  };
+
+  const handleSave = () => {
+    props.updateVisualmetadataEntity({
+      viewId,
+      visualMetadata: props.visualmetadataEntity,
+    });
+    VisualMetadataContainerUpdate(props.visualmetadataEntity.id, props.visualmetadataEntity, 'id');
+    handleClose();
+  };
+  useEffect(() => {
+    if (visualizationId) {
+      props.getVisualmetadataEntity(visualizationId);
+      props.getfeatureEntities(viewId);
+      props.getViewEntity(viewId);
+    }
+  }, []);
+
+  const onExchangeMetadata = data => {
+    const metaData = data.body === '' ? { data: [] } : JSON.parse(data.body);
+    setData(metaData.data);
+    if (data.headers.request === 'filters') {
+       // console.log('filter data');
+    } else {
+      renderVisualization(props.visualmetadataEntity, metaData.data, 'visualization-edit');
+    }
+  };
+
+  useEffect(() => {
+    setVisual(props.visualmetadataEntity);
+    if (props.visualmetadataEntity.fields && ValidateFields(props.visualmetadataEntity.fields)) {
+      subscribeWebSocket('/user/exchange/metaData/' + props.visualmetadataEntity.id, onExchangeMetadata);
+      const visualMetadata = VisualWrap(props.visualmetadataEntity);
+      const queryDTO = visualMetadata.getQueryParameters(props.visualmetadataEntity, null, null, null);
+      const body = {
+        queryDTO,
+        visualMetadata,
+        validationType: 'REQUIRED_FIELDS',
+        actionType: null,
+        type: 'share-link',
+      };
+      forwardCall(props.view?.viewDashboard?.dashboardDatasource?.id, body, props.view.id);
+    }
+  }, [props.visualmetadataEntity, getData]);
```

Comment:
why getData function is part of useEffect props

Reply:
done